game/src/edit/new_traffic_signals.rs

```diff
@@ -0,0 +1,766 @@
+            let mut new = make_side_panel(ctx, app, &self.members, self.current_phase);
+            new.restore(ctx, &self.side_panel);
+            self.side_panel = new;
+        } else {
+            self.current_phase = idx;
+            self.side_panel = make_side_panel(ctx, app, &self.members, self.current_phase);
+            // TODO Maybe center of previous member
+            self.side_panel
+                .scroll_to_member(ctx, format!("phase {}", idx + 1));
+        }
+
+        self.draw_current = self.recalc_draw_current(ctx, app);
+    }
+
+    fn recalc_draw_current(&self, ctx: &mut EventCtx, app: &App) -> Drawable {
+        let mut batch = GeomBatch::new();
+
+        for i in &self.members {
+            let signal = app.primary.map.get_traffic_signal(*i);
+            let mut phase = signal.phases[self.current_phase].clone();
+            if let Some((id, _)) = self.group_selected {
+                if id.parent == signal.id {
+                    phase.edit_group(&signal.turn_groups[&id], TurnPriority::Banned);
+                }
+            }
+            draw_signal_phase(
+                ctx.prerender,
+                &phase,
+                signal.id,
+                None,
+                &mut batch,
+                app,
+                app.opts.traffic_signal_style.clone(),
+            );
+        }
+
+        for tg in &self.groups {
+            let signal = app.primary.map.get_traffic_signal(tg.id.parent);
+            if self
+                .group_selected
+                .as_ref()
+                .map(|(id, _)| *id == tg.id)
+                .unwrap_or(false)
+            {
+                draw_selected_group(
+                    app,
+                    &mut batch,
+                    tg,
+                    &signal.turn_groups[&tg.id],
+                    self.group_selected.unwrap().1,
+                );
+            } else {
+                batch.push(app.cs.signal_turn_block_bg, tg.block.clone());
+                let phase = &signal.phases[self.current_phase];
+                let arrow_color = match phase.get_priority_of_group(tg.id) {
+                    TurnPriority::Protected => app.cs.signal_protected_turn,
+                    TurnPriority::Yield => app.cs.signal_permitted_turn,
+                    TurnPriority::Banned => app.cs.signal_banned_turn,
+                };
+                batch.push(arrow_color, tg.arrow.clone());
+            }
+        }
+        ctx.upload(batch)
+    }
+}
+
+impl State for NewTrafficSignalEditor {
+    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
+        ctx.canvas_movement();
+
+        let canonical_signal = app
+            .primary
+            .map
+            .get_traffic_signal(*self.members.iter().next().unwrap());
+        let num_phases = canonical_signal.phases.len();
+
+        match self.side_panel.event(ctx) {
+            Outcome::Clicked(x) => {
+                if x == "Add new phase" {
```

Comment:
I'd never actually used the traffic phase editor - the lanes/turns enable/disable UI is nice!

Reply:
Thanks! It's gone through lots of iteration with Yuwen already, and I suspect we have a long way to go. I think signal editing is likely the most complex UI problem abst tries to tackle.